### PR TITLE
Kamu UI 406 detect when dataset cache is stale

### PR DESCRIPTION
**Pull request by dmitriy-borzenko**
Closes: https://github.com/kamu-data/kamu-web-ui/issues/406

### CODE REVIEW

File: src/app/api/dataset-flow.api.ts [hunk 1]
```diff
@@ -53,21 +54,35 @@ export class DatasetFlowApi {
...
-        );
+        return this.datasetTriggerFlowGQL
+            .mutate(
+                { ...params },
+                {
+                    update: (cache) => {
+                        updateCacheHelper(cache, {
+                            accountId: params.accountId,
+                            datasetId: params.datasetId,
+                            fieldNames: ["data"],
```
Comment (zaychenko-sergei) [inline src/app/api/dataset-flow.api.ts:70]: Why just `data`? Manual trigger of update or compacting/reset will most likely change the HEAD reference too, and that affects `metadata` as well. See we are resetting `metadata` when committing events in other places.

File: src/app/api/dataset-flow.api.ts [hunk 2]
```diff
@@ -84,15 +99,29 @@ export class DatasetFlowApi {
...
+                {
+                    ...params,
+                },
+                {
+                    update: (cache) => {
+                        if (!params.paused) {
+                            updateCacheHelper(cache, {
+                                accountId: params.accountId,
+                                datasetId: params.datasetId,
+                                fieldNames: ["data"],
```
Comment (zaychenko-sergei) [inline src/app/api/dataset-flow.api.ts:119]: Same here

File: src/app/api/dataset-flow.api.ts [hunk 3]
```diff
@@ -107,15 +136,29 @@ export class DatasetFlowApi {
...
```
Comment (zaychenko-sergei) [inline src/app/api/dataset-flow.api.ts:156]: Same here

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
Not sure this is going to help much.
Imagine the following scenario:
 - user clicks manual trigger or one of the config updates
 - the flow & tasks are submitted, but have to wait for another long operation already running
 - there might be a few minutes before it completes
 - okay, you've reset `data` (and `metadata` after fixes), but as the corresponding task awaits for a while, there will be another update you will miss.
 
 I think we should rather consider how we are going to implement real-time updates.

**Comment by zaychenko-sergei** [comment]
Let's also consider checking this idea:

image

I'm not sure because of possible performance impact, but we can try.
So you need to make a little uncached query about dataset's HEAD, and do this on every route event when you are navigating with the dataset or it's sub-pages.

The "last known"  hash is already available in Apollo cache, you don't ahve to save it elsewhere, I suppose.

File: src/app/api/dataset-flow.api.ts (hunk 3 above)
Comment (dmitriy-borzenko) [reply src/app/api/dataset-flow.api.ts]: Removed

File: src/app/api/dataset-flow.api.ts (hunk 1 above)
Comment (dmitriy-borzenko) [reply src/app/api/dataset-flow.api.ts]: removed

File: CHANGELOG.md [hunk 4]
```diff
@@ -16,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [0.26.1] - 2024-09-06
 ### Fixed
 - Set consistent state with `force update` link
+- Updated overview tab in real time when the flow is finished
```
Comment (zaychenko-sergei) [inline CHANGELOG.md:21]: Merge of history to published release?

File: src/app/api/dataset.api.ts [hunk 5]
```diff
@@ -439,6 +441,25 @@ export class DatasetApi {
             );
     }
 
+    public datasetHashLastBlock(accountName: string, datasetName: string): Observable<DatasetHashLastBlockQuery> {
```
Comment (zaychenko-sergei) [inline src/app/api/dataset.api.ts:444]: Use term "head" everywhere. What is "first" and what is "last" in blockchain is very ambigious.

File: src/app/api/gql/dataset-hash-last-block.graphql [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  datasets {
+    byOwnerAndName(accountName: $accountName, datasetName: $datasetName) {
+      metadata {
+        chain {
+          blocks(perPage: 1, page: 0) {
+            nodes {
+              blockHash
+            }
+          }
+        }
```
Comment (zaychenko-sergei) [inline src/app/api/gql/dataset-hash-last-block.graphql:11]:
```
type MetadataChain {
	"""
	Returns all named metadata block references
	"""
	refs: [BlockRef!]!
```

You need a ref named "head".

File: src/app/dataset-view/dataset.component.ts [hunk 7]
```diff
@@ -63,9 +77,34 @@ export class DatasetComponent extends BaseDatasetDataComponent implements OnInit
...
+                .pipe(
+                    map((hash: string) => {
+                        if (hash !== this.currentHashLastBlock && this.datasetBasics) {
+                            this.updateCache(this.datasetBasics);
+                        }
+                    }),
+                    switchMap(() => this.datasetService.requestDatasetMainData(urlDatasetInfo)),
+                    tap(() => {
+                        this.mainDatasetQueryComplete$.next(urlDatasetInfo);
+                    }),
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/dataset.component.ts:93]: Isn't it enough just to clean the cache?

File: CHANGELOG.md (hunk 4 above)
Comment (dmitriy-borzenko) [reply CHANGELOG.md:21]: Removed, it's not true

File: src/app/dataset-view/dataset.component.ts (hunk 7 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/dataset.component.ts:93]: No, need to make a main request too

File: src/app/dataset-view/dataset.component.ts [hunk 8]
```diff
@@ -55,6 +59,16 @@ export class DatasetComponent extends BaseDatasetDataComponent implements OnInit
...
+        this.fetchHashHead(urlDatasetInfo);
+    }
+
+    private fetchHashHead(datasetInfo: DatasetInfo): void {
+        this.datasetService
+            .requestDatasetHeadBlockHash(datasetInfo.accountName, datasetInfo.datasetName)
+            .pipe(takeUntilDestroyed(this.destroyRef))
+            .subscribe((blockHash: string) => {
+                this.currentHeadBlockHash = blockHash;
+            });
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/dataset.component.ts:71]: This does not look like presentation level logic to me, should be in some service.

File: src/app/dataset-view/dataset.component.ts [hunk 9]
```diff
@@ -63,9 +77,34 @@ export class DatasetComponent extends BaseDatasetDataComponent implements OnInit
...
+    }
+
+    private updateCache(datasetBasics: DatasetBasicsFragment): void {
+        const cache = this.injector.get(APOLLO_OPTIONS).cache;
+        if (cache) {
+            updateCacheHelper(cache, {
+                accountId: datasetBasics.owner.id,
+                datasetId: datasetBasics.id,
+                fieldNames: ["metadata", "data"],
+            });
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/dataset.component.ts:107]: This isn't presentation too

File: src/app/dataset-view/dataset.component.ts [hunk 10]
```diff
@@ -55,6 +59,16 @@ export class DatasetComponent extends BaseDatasetDataComponent implements OnInit
             });
 
         this.datasetPermissions$ = this.datasetSubsService.permissionsChanges;
+        this.fetchHashHead(urlDatasetInfo);
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/dataset.component.ts:62]: Do we have to make this query if cache is currently empty?

File: src/app/dataset-view/dataset.component.ts [hunk 11]
```diff
@@ -63,9 +77,34 @@ export class DatasetComponent extends BaseDatasetDataComponent implements OnInit
             _.isNil(this.datasetBasics) ||
             this.datasetBasics.name !== urlDatasetInfo.datasetName ||
             this.datasetBasics.owner.accountName !== urlDatasetInfo.accountName ||
-            this.datasetViewType === DatasetViewTypeEnum.Flows
+            [DatasetViewTypeEnum.Overview, DatasetViewTypeEnum.Metadata].includes(this.datasetViewType)
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/dataset.component.ts:80]: Why only these tabs? When I select `Data` tab, I want to see relevant schema too.

File: src/app/dataset-view/dataset.component.ts [hunk 12]
```diff
@@ -63,9 +77,34 @@ export class DatasetComponent extends BaseDatasetDataComponent implements OnInit
...
+                    }),
+                    switchMap(() => this.datasetService.requestDatasetMainData(urlDatasetInfo)),
+                    tap(() => {
+                        this.mainDatasetQueryComplete$.next(urlDatasetInfo);
+                    }),
+                    takeUntilDestroyed(this.destroyRef),
+                )
+                .subscribe();
+        }
+    }
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/dataset.component.ts:98]: I can see that everything is re-queried after returning from History tab.
1) Open dataset Overview initially
2) Switch to History
3) Switch to Overview back - I can see main data is re-queried without any reason.

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
With this approach, do we still need the following previous actions:

1) `DatasetService` has `updatePage` method, which forces refresh after explicit commit events:

```
    private updatePage(accountName: string, datasetName: string): void {
        this.datasetService
            .requestDatasetMainData({
                accountName,
                datasetName,
            })
            .subscribe();

        this.navigationService.navigateToDatasetView({
            accountName,
            datasetName,
            tab: DatasetViewTypeEnum.Overview,
        });
    }
```

This looks redundand now. We should somehow re-trigger the new update process.

2) Why do we need things like cache update now, if the view is auto-refreshed:

```
    private updateCache(datasetBasics: DatasetBasicsFragment): void {
        const cache = this.injector.get(APOLLO_OPTIONS).cache;
        if (cache) {
            updateCacheHelper(cache, {
                accountId: datasetBasics.owner.id,
                datasetId: datasetBasics.id,
                fieldNames: ["metadata", "data"],
            });
        }
    }
```

File: src/app/dataset-view/dataset.component.ts (hunk 12 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/dataset.component.ts:98]: I can't reproduce it, please check again

File: src/app/dataset-view/dataset.component.ts (hunk 11 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/dataset.component.ts:80]: Added `Data` tab

File: src/app/dataset-view/dataset.component.ts (hunk 8 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/dataset.component.ts]: Moved to service

**Comment by dmitriy-borzenko** [comment]
> With this approach, do we still need the following previous actions:
> 
> 1. `DatasetService` has `updatePage` method, which forces refresh after explicit commit events:
> 
> ```
>     private updatePage(accountName: string, datasetName: string): void {
>         this.datasetService
>             .requestDatasetMainData({
>                 accountName,
>                 datasetName,
>             })
>             .subscribe();
> 
>         this.navigationService.navigateToDatasetView({
>             accountName,
>             datasetName,
>             tab: DatasetViewTypeEnum.Overview,
>         });
>     }
> ```
> 
> This looks redundand now. We should somehow re-trigger the new update process.
> 
> 2. Why do we need things like cache update now, if the view is auto-refreshed:
> 
> ```
>     private updateCache(datasetBasics: DatasetBasicsFragment): void {
>         const cache = this.injector.get(APOLLO_OPTIONS).cache;
>         if (cache) {
>             updateCacheHelper(cache, {
>                 accountId: datasetBasics.owner.id,
>                 datasetId: datasetBasics.id,
>                 fieldNames: ["metadata", "data"],
>             });
>         }
>     }
> ```

1. Deteted request main data, but navigation is needed for the router subscription to work
2. Without updating cache the page does not update

File: src/app/dataset-view/dataset.component.ts (hunk 12 above)
Comment (zaychenko-sergei) [reply src/app/dataset-view/dataset.component.ts]: Can't reproduce it now too, weird.

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
Thoughts after new round:

1. When we detect the hash has changed, we have no idea what exactly caused it, so should wipe the entire dataset record in the cache, not just particular fields.

2. Many mutations where we currently touch the cache (commitEvent, updateReadme, setWatermark, uploading a fle), probably do not need this cache reset, as they modify the HEAD, and the default logic should work, assuming we trigger the navigation or refresh.

3. Some mutations do not change the head, so they need cache cleanup as currently (operations with env vars, also renaming dataset).

4. File upload triggers yet another re-request of main data, which is redundand.

```
  private updatePage(datasetBasics: DatasetBasicsFragment): void {
        this.updateCache(datasetBasics);
         // DO WE NEED THIS REQUEST NOW?
        this.datasetService
            .requestDatasetMainData({
                accountName: datasetBasics.owner.accountName,
                datasetName: datasetBasics.name,
            })
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe(),
            this.navigationService.navigateToDatasetView({
                accountName: datasetBasics.owner.accountName,
                datasetName: datasetBasics.name,
                tab: DatasetViewTypeEnum.Overview,
            });
    }
    ```

**Comment by dmitriy-borzenko** [comment]
> Thoughts after new round:
> 
> 1. When we detect the hash has changed, we have no idea what exactly caused it, so should wipe the entire dataset record in the cache, not just particular fields.
> 2. Many mutations where we currently touch the cache (commitEvent, updateReadme, setWatermark, uploading a fle), probably do not need this cache reset, as they modify the HEAD, and the default logic should work, assuming we trigger the navigation or refresh.
> 3. Some mutations do not change the head, so they need cache cleanup as currently (operations with env vars, also renaming dataset).
> 4. File upload triggers yet another re-request of main data, which is redundand.
> 
> ```
>   private updatePage(datasetBasics: DatasetBasicsFragment): void {
>         this.updateCache(datasetBasics);
>          // DO WE NEED THIS REQUEST NOW?
>         this.datasetService
>             .requestDatasetMainData({
>                 accountName: datasetBasics.owner.accountName,
>                 datasetName: datasetBasics.name,
>             })
>             .pipe(takeUntilDestroyed(this.destroyRef))
>             .subscribe(),
>             this.navigationService.navigateToDatasetView({
>                 accountName: datasetBasics.owner.accountName,
>                 datasetName: datasetBasics.name,
>                 tab: DatasetViewTypeEnum.Overview,
>             });
>     }
>     ```
> ```

All comments have been corrected

**Review by zaychenko-sergei — APPROVED** [review]

**Pull request merged**